src/output-json-alert.c

```diff
@@ -169,6 +171,32 @@ static void AlertJsonSourceTarget(const Packet *p, const PacketAlert *pa,
     jb_close(js);
 }
 
+static void AlertJsonReference(
+        AlertJsonOutputCtx *json_output_ctx, const PacketAlert *pa, JsonBuilder *jb)
+{
+    if (!pa->s->references) {
+        return;
+    }
+
+    const DetectReference *kv = pa->s->references;
+    jb_open_array(jb, "references");
+    while (kv) {
+        BUG_ON(kv->key_len > REFERENCE_SYSTEM_NAME_MAX);
+        BUG_ON(kv->reference_len > REFERENCE_CONTENT_NAME_MAX);
+        const size_t size_needed = kv->key_len + kv->reference_len + 1;
+        char kv_store[size_needed];
+        // Some reference values may contain the scheme (unnecessarily, but
+        // allowed) so check for that here.
+        if (kv->reference_len >= kv->key_len && strncmp(kv->reference, kv->key, kv->key_len) == 0)
+            snprintf(kv_store, size_needed, "%s", kv->reference);
+        else
+            snprintf(kv_store, size_needed, "%s%s", kv->key, kv->reference);
```

Comment:
Would it make more sense to sanitize the references when loaded, rather than every time an alert is triggered?

Reply:
Good idea

Reply:
The current regex for parsing `reference` keyword values will parse `reference=url,https://www.anysite.com` as `url` and `https`. 

`url` expands to `https://`

Should we presume that if the content is `https` (because of the regex) and that's a substring of the URL value, then the rule be flagged as invalid?

Reply:
I don't quite follow. This looks valid to me.  If the value contains a scheme, I'd use that over the scheme provided as the first parameter. Does that make sense?

Reply:
That makes sense and is possible with additional parsing support.

Current parsing will create this reference structure (references are never included in an output log so its not noticed):
```
key = https://
reference=https
 ```
 
 We can change how reference keywords are parsed and flag references that use `url` and the reference value (incorrectly) contains a scheme.
 
 E.g., `reference=url,FTP://ftp.google.com/some/article` would overwrite the scheme in `reference.config`'s URL (`https://`
 ```
 key=ftp://
 reference=ftp.google.com/some/article
 ```
 
 Is this what you're asking for?